### PR TITLE
Feature/slim 1993 null data

### PR DESCRIPTION
**Pull request by coendamen**

### CODE REVIEW

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java [hunk 1]
```diff
@@ -110,43 +163,74 @@ public void testHappy() throws Exception {
...
         when(cosemDateTime.asDateTime()).thenReturn(bufferedDateTime);
 
-        // TODO: mock other calls
-
-        final PeriodicMeterReadsRequestDto request = new PeriodicMeterReadsRequestDto(periodType, new Date(from),
-                new Date(to));
+        final DlmsMeterValueDto meterValue1 = mock(DlmsMeterValueDto.class);
+        final DlmsMeterValueDto meterValue2 = mock(DlmsMeterValueDto.class);
+        when(this.dlmsHelper.getScaledMeterValue(data1, null, "positiveActiveEnergyTariff1")).thenReturn(meterValue1);
+        when(this.dlmsHelper.getScaledMeterValue(data4, null, "positiveActiveEnergyTariff2")).thenReturn(meterValue2);
```
Comment (stefanermens) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java]: Hier kan je alle waardes mocken:
        when(this.dlmsHelper.getScaledMeterValue(data1, null, "positiveActiveEnergyTariff1")).thenReturn(meterValue1);
        when(this.dlmsHelper.getScaledMeterValue(data2, null, "positiveActiveEnergyTariff2")).thenReturn(meterValue2);
        when(this.dlmsHelper.getScaledMeterValue(data3, null, "negativeActiveEnergyTariff1")).thenReturn(meterValue3);
        when(this.dlmsHelper.getScaledMeterValue(data4, null, "negativeActiveEnergyTariff2")).thenReturn(meterValue4);

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java [hunk 2]
```diff
@@ -110,43 +163,74 @@ public void testHappy() throws Exception {
...
+                any(DlmsObjectType.class), any(Medium.class));
+
+        // ASSERT - the result should contain 2 values
         final List<PeriodicMeterReadsResponseItemDto> periodicMeterReads = result.getPeriodicMeterReads();
-        // TODO: assert contents of result
+
+        assertThat(periodicMeterReads.size()).isEqualTo(2);
+
+        periodicMeterReads.forEach(p -> {
+            assertThat(p.getLogTime()).isNotNull();
```
Comment (stefanermens) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java]: En dan kan je hier controleren of de waardes goed gebruikt zijn:
            assertThat(p.getActiveEnergyImportTariffOne()).isEqualTo(meterValue1);
            assertThat(p.getActiveEnergyImportTariffTwo()).isEqualTo(meterValue2);
            assertThat(p.getActiveEnergyExportTariffOne()).isEqualTo(meterValue3);
            assertThat(p.getActiveEnergyExportTariffTwo()).isEqualTo(meterValue4);

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java [hunk 3]
```diff
@@ -101,7 +153,8 @@ public void testHappy() throws Exception {
         final DataObject bufferedObject = mock(DataObject.class);
         when(bufferedObject.getValue()).thenReturn(asList(data0, data1, data2, data3, data4, data5));
         final DataObject resultData = mock(DataObject.class);
-        when(resultData.getValue()).thenReturn(Collections.singletonList(bufferedObject));
+        when(resultData.getValue()).thenReturn(Arrays.asList(bufferedObject, bufferedObject));
```
Comment (stefanermens) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java]: Je gebruikt 2 x bufferedObject. Mooier zou zijn om 2 verschillende bufferedObjects aan te maken. Het kost wat meer mock-werk, maar dan test je ook of de code bv niet per ongeluk altijd alleen maar de eerste waarde uit de lijst haalt.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsGasCommandExecutor.java [hunk 4]
```diff
@@ -150,12 +141,17 @@ public PeriodicMeterReadGasResponseDto execute(final DlmsConnectionManager conn,
     }
 
     private PeriodicMeterReadsGasResponseItemDto convertToResponseItem(
```
Comment (jenkins-ip-10-4-24-184) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsGasCommandExecutor.java]: MAJOR Method has 7 parameters, which is greater than 6 authorized. [rule](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS00107)

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/utils/DlmsHelper.java [hunk 5]
```diff
@@ -886,13 +886,13 @@ private void logAndThrowExceptionForUnexpectedResultData(final DataObject result
     public void validateBufferedDateTime(final DateTime bufferedDateTime, final CosemDateTimeDto cosemDateTime,
```
Comment (jenkins-ip-10-4-24-184) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/utils/DlmsHelper.java]: MAJOR Remove this unused method parameter "cosemDateTime". [rule](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1172)

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/utils/DlmsHelper.java [hunk 6]
```diff
@@ -886,13 +886,13 @@ private void logAndThrowExceptionForUnexpectedResultData(final DataObject result
...
             final DateTime beginDateTime, final DateTime endDateTime) throws BufferedDateTimeValidationException {
 
-        if (bufferedDateTime == null) {
-            final DateTimeFormatter dtf = ISODateTimeFormat.dateTime();
-            throw new BufferedDateTimeValidationException(
-                    "Not using an object from capture buffer (clock=" + cosemDateTime
-                            + "), because the date does not match the given period, since it is not fully specified: ["
-                            + dtf.print(beginDateTime) + " .. " + dtf.print(endDateTime) + "].");
-        }
+//        if (bufferedDateTime == null) {
```
Comment (jenkins-ip-10-4-24-184) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/utils/DlmsHelper.java]: MAJOR This block of commented-out lines of code should be removed. [rule](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/utils/DlmsHelper.java [hunk 7]
```diff
@@ -886,13 +886,13 @@ private void logAndThrowExceptionForUnexpectedResultData(final DataObject result
...
+//            final DateTimeFormatter dtf = ISODateTimeFormat.dateTime();
+//            throw new BufferedDateTimeValidationException(
+//                    "Not using an object from capture buffer (clock=" + cosemDateTime
+//                            + "), because the date does not match the given period, since it is not fully specified: ["
+//                            + dtf.print(beginDateTime) + " .. " + dtf.print(endDateTime) + "].");
```
Comment (jenkins-ip-10-4-24-184) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/utils/DlmsHelper.java]: MAJOR This block of commented-out lines of code should be removed. [rule](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutor.java [hunk 8]
```diff
@@ -90,160 +71,207 @@ public PeriodicMeterReadsRequestDto fromBundleRequestInput(final ActionRequestDt
...
-        switch (periodType) {
-        case INTERVAL:
-            return this.getNextPeriodicMeterReadsForInterval(bufferedObjects, bufferedDateTime, results);
-        case DAILY:
-            return this.getNextPeriodicMeterReadsForDaily(bufferedObjects, bufferedDateTime, results);
-        case MONTHLY:
-            return this.getNextPeriodicMeterReadsForMonthly(bufferedObjects, bufferedDateTime, results);
-        default:
-            throw new AssertionError("Unknown PeriodType: " + periodType);
+    private PeriodicMeterReadsResponseItemDto convertToResponseItem(
```
Comment (jenkins-ip-10-4-24-184) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutor.java]: MAJOR Method has 7 parameters, which is greater than 6 authorized. [rule](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS00107)

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/AbstractPeriodicMeterReadsCommandExecutor.java [hunk 9]
```diff
@@ -1,20 +1,196 @@
...
+
+        switch (periodTypeDto) {
+            case DAILY:
+                return Date.from(previousLogTime.toInstant().plus(Duration.ofDays(1)));
+            case MONTHLY:
+                LocalDateTime localDateTime = LocalDateTime.ofInstant(previousLogTime.toInstant(), ZoneId.systemDefault()).plusMonths(1);
+
+                return Date.from(localDateTime.atZone(ZoneId.systemDefault())
+                        .toInstant());
+            case INTERVAL:
```
Comment (jenkins-ip-10-4-24-184) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/AbstractPeriodicMeterReadsCommandExecutor.java]: MAJOR Reduce this switch case number of lines from 7 to at most 5, for example by extracting code into methods. [rule](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1151)

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java (hunk 1 above)
Comment (coendamen) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java]: done

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java (hunk 2 above)
Comment (coendamen) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java]: done

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java (hunk 3 above)
Comment (coendamen) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java]: done, created one different mock, and asserted 2 values.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsGasCommandExecutor.java (hunk 4 above)
Comment (coendamen) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsGasCommandExecutor.java]: done

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/utils/DlmsHelper.java (hunk 5 above)
Comment (coendamen) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/utils/DlmsHelper.java]: done

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/utils/DlmsHelper.java (hunk 6 above)
Comment (coendamen) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/utils/DlmsHelper.java]: done

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/utils/DlmsHelper.java (hunk 7 above)
Comment (coendamen) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/utils/DlmsHelper.java]: done

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/AbstractPeriodicMeterReadsCommandExecutor.java (hunk 9 above)
Comment (coendamen) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/AbstractPeriodicMeterReadsCommandExecutor.java]: done

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java (hunk 1 above)
Comment (akoster) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java]: This violates the 'Law of Demeter', see the famous post: https://tuhrig.de/everytime-a-mock-returns-a-mock-a-fairy-dies/

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,537 @@
...
+
+    private Date TIME_FROM;
+    private Date TIME_TO;
+    private DataObject PERIOD_1_CLOCK;
+    private DataObject PERIOD_2_CLOCK;
+    private Date PERIOD_1_CLOCK_VALUE;
+    private Date PERIOD_2_CLOCK_VALUE;
+    private Date PERIOD_2_CLOCK_VALUE_NULL_DATA_PERIOD_15MIN;
+    private Date PERIOD_2_CLOCK_VALUE_NULL_DATA_PERIOD_DAILY;
+    private Date PERIOD_2_CLOCK_VALUE_NULL_DATA_PERIOD_MONTHLY;
```
Comment (akoster) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorIntegrationTest.java]: The above fields are not constants so should be named with camel case, not uppercase-with-underscores

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,537 @@
...
+import org.opensmartgridplatform.dto.valueobjects.smartmetering.PeriodicMeterReadsResponseDto;
+import org.opensmartgridplatform.dto.valueobjects.smartmetering.PeriodicMeterReadsResponseItemDto;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@RunWith(MockitoJUnitRunner.class)
+public class GetPeriodicMeterReadsCommandExecutorIntegrationTest {
```
Comment (akoster) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorIntegrationTest.java]: The name 'IntegrationTest' is confusing here because this test is run as a unit test by the maven-surefire-plugin. We also use the maven-failsafe-plugin which runs tests matching: `<include>%regex[.*integrationtests.*class]</include>` in the `integration-test` phase.

I think you named it such because this test is not a pure 'mockist' style unit tests. There is a 'classical' view on unit testing which would call this a real unit test, where the unit happens to be larger than a single class. I don't prefer such tests myself since the tests tend to become hard to read, repetitive and prone to miss edge cases. Other developers (read: @stefanermens) however are of a different opinion.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java [hunk 12]
```diff
@@ -57,51 +54,108 @@
...
         when(this.connectionManager.getDlmsMessageListener()).thenReturn(this.dlmsMessageListener);
     }
 
+    @Test
+    public void testExecuteNullRequest() throws ProtocolAdapterException {
+        try {
+            this.executor.execute(this.connectionManager, this.device, null);
+            fail("Calling execute with null query should fail");
+        } catch (final IllegalArgumentException e) {
+            // expected exception
```
Comment (akoster) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java]: You can specify expected exceptions using `@Test(expected = IllegalArgumentException.class)`. You can also remove the `fail(..)` statement in that case.

For more fine grained controle you could use `org.junit.rules.ExpectedException`

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java [hunk 13]
```diff
@@ -110,43 +164,82 @@ public void testHappy() throws Exception {
...
+
+        periodicMeterReads.forEach(p -> {
+            assertThat(p.getLogTime()).isNotNull();
+            assertThat(p.getActiveEnergyImportTariffOne()).isEqualTo(meterValue1);
+            assertThat(p.getActiveEnergyImportTariffTwo()).isEqualTo(meterValue2);
+            assertThat(p.getActiveEnergyExportTariffOne()).isEqualTo(meterValue3);
+            assertThat(p.getActiveEnergyExportTariffTwo()).isEqualTo(meterValue4);
+        });
+
+    }
```
Comment (akoster) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java]: The method `testHappy` is very long. I suggest extracting setup logic as methods to make it more readable.

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledGetPeriodicMeterReadsSteps.java [hunk 14]
```diff
@@ -44,6 +46,10 @@ public void theBundleResponseShouldContainAGetPeriodicMeterReadsResponse() throw
 
         final Response response = this.getNextBundleResponse();
 
-        assertTrue("Not a valid response", response instanceof PeriodicMeterReadsResponse);
+        LOGGER.info("Got PeriodicMeterReadResponse. exception == {}, result == {}", response.getException(), response.getResultString());
+
+        String errorMessage = "Not a valid response " + response.getException() + " -- result -- " + response.getResultString();
+
+       Assert.assertTrue(errorMessage, response instanceof PeriodicMeterReadsResponse);
```
[the remark this answers is not in the document]
Comment (coendamen) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledGetPeriodicMeterReadsSteps.java]: this code is to debug the failing cucumber test, so will be removed once it is clear why the test is failing, Or when the test is removed.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/AbstractPeriodicMeterReadsCommandExecutor.java [hunk 15]
```diff
@@ -1,20 +1,217 @@
...
+     * @param intervalTime    the interval time for this device to be taken into account when the periodTypeDto is INTERVAL
+     * @return the derived date based on the previous meter read record, or null if it cannot be determined
+     */
+    private Date calculateIntervalDate(final PeriodTypeDto periodTypeDto,
+                                       final Date previousLogTime,
+                                       final ProfileCaptureTime intervalTime) throws BufferedDateTimeValidationException {
+
+        if (previousLogTime == null) {
+            return null;
+        }
```
[the remark this answers is not in the document]
Comment (coendamen) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/AbstractPeriodicMeterReadsCommandExecutor.java]: I understood that the first record always returns a Date. So all subsequent calculations would return a Date. ReadClock is supposed to never return null. It is supposed to fail in that case. 
we could agree that in that case, a BufferedDateTimeValidationException should be thrown

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/AbstractPeriodicMeterReadsCommandExecutor.java [hunk 16]
```diff
@@ -1,20 +1,217 @@
...
+
+        if (dlmsObject instanceof DlmsProfile) {
+            DlmsProfile profile = (DlmsProfile) dlmsObject;
+
+            getLogger().info("Capture time of this device is {} ", profile.getCaptureTime());
+            return profile.getCaptureTime();
+        }
+
+        return null;
+    }
```
[the remark this answers is not in the document]
Comment (coendamen) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/AbstractPeriodicMeterReadsCommandExecutor.java]: it is possible to rewrite it, but this is just another way of writing the logic. can you explain why the current code is less readable ?

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/AbstractPeriodicMeterReadsCommandExecutor.java [hunk 17]
```diff
@@ -1,20 +1,217 @@
...
+     * @return the date of the buffered {@link DataObject} or null if it cannot be determined
+     * @throws ProtocolAdapterException
+     * @throws BufferedDateTimeValidationException in case the date is invalid or null
+     */
+    Date readClock(final ConversionContext ctx,
+                   final Date previousLogTime,
+                   final DlmsHelper dlmsHelper)
+            throws ProtocolAdapterException, BufferedDateTimeValidationException {
+
+        final PeriodTypeDto queryPeriodType = ctx.periodicMeterReadsQuery.getPeriodType();
```
[the remark this answers is not in the document]
Comment (coendamen) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/AbstractPeriodicMeterReadsCommandExecutor.java]: the readclock method follows two paths, either it gets the date (I did not touch that logic) or it calculates the date based on the previous value (the fix). The above changes suggests adding several new methods and changing it to a streaming API. We could argue whether that is really necessary.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java (hunk 12 above)
Comment (coendamen) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java]: I know, I inherited this test and although it can be improved, it can be argued that this is part of this story.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorIntegrationTest.java (hunk 11 above)
Comment (coendamen) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorIntegrationTest.java]: I inherited this setup and did not touch the structure.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorIntegrationTest.java (hunk 10 above)
Comment (coendamen) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorIntegrationTest.java]: you are right, I forgot to rename them after refactoring the test. Also, this should be picked up by Sonar.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java (hunk 13 above)
Comment (coendamen) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorTest.java]: could be done in another story.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutor.java [hunk 18]
```diff
@@ -90,160 +71,203 @@ public PeriodicMeterReadsRequestDto fromBundleRequestInput(final ActionRequestDt
...
+        // Get scaler unit from result list. Note: "index + 1" because the first result is the array with values
+        // and should be skipped. The first scaler unit is at index 1.
+        if (scalerUnitIndex != null) {
+            return getResultList.get(scalerUnitIndex + 1).getResultData();
         }
 
-        final Set<AmrProfileStatusCodeFlagDto> flags = this.amrProfileStatusCodeHelper
-                .toAmrProfileStatusCodeFlags(amrProfileStatusData.getValue());
-        return new AmrProfileStatusCodeDto(flags);
+        return null;
```
[the remark this answers is not in the document]
Comment (coendamen) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutor.java]: you are right, I did not touch this code, only reshuffled it. but I can improve that in this PR

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledGetPeriodicMeterReadsSteps.java [hunk 19]
```diff
@@ -1,26 +1,28 @@
 /**
  * Copyright 2017 Smart Society Services B.V.
- *
+ * <p>
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.  You may obtain a copy of the License at
- *
- *     http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
```
[the remark this answers is not in the document]
Comment (coendamen) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledGetPeriodicMeterReadsSteps.java]: you are right, I will correct this

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutor.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (coendamen) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutor.java]: you are right, although returning null is not necessarily a code smell, it can be improved with an Optional. I will fix that

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorIntegrationTest.java (hunk 11 above)
Comment (stefanermens) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorIntegrationTest.java]: The main reason for this test is to verify that when all the classes work together, the correct DLMS request is created. This is essential to test since a single small difference could cause the smart meter to send a wrong reply or no reply at all. This test tests all possible requests from the SMHE and compares the created requests with known good requests, which have been tested with real meters. So with a single test we can see that with a certain input (e.g. Daily meter reads, gas meter, SMR5) we get the right request. 

With pure 'mockist' style unit tests it would be very hard to prove the same. And when a code change is needed, the 'mockist' style unit test would need to be changed as well, which might lead to a wrong request being created, but not detected by the test.

About the naming we could have endless discussions. We did have a discussion last time and then choose to name it IntegrationTests. I don't mind what we choose, as long as we don't discuss it over and over again.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/AbstractPeriodicMeterReadsCommandExecutor.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (akoster) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/AbstractPeriodicMeterReadsCommandExecutor.java]: Because you are mixing new declarative style with traditional control flow.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/AbstractPeriodicMeterReadsCommandExecutor.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (akoster) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/AbstractPeriodicMeterReadsCommandExecutor.java]: The existing code is not clean because the methods are way to large and do multiple things. I am suggesting ways to make the code cleaner. Do with that what you like.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutor.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (akoster) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutor.java]: I disagree. Returning null is definitely a code smell. Support for example: https://blog.jetbrains.com/idea/2017/08/code-smells-null/
